src/compiler/binder.ts

```diff
@@ -3138,9 +3137,18 @@ namespace ts {
                 transformFlags |= TransformFlags.AssertES2015;
                 break;
 
+            case SyntaxKind.ForOfStatement:
+                // This node is either ES2015 syntax or ES2017 syntax (if it is a for-await-of).
+                if ((<ForOfStatement>node).awaitModifier) {
+                    transformFlags |= TransformFlags.AssertESNext;

```

Comment:
The way that the esnext transformer works now is to always descend through the tree if *some* node is marked `ESNext`. The `SpreadAssignment` case handles that further down, so we can reduce false positives in `for..of` triggering the esnext transform when it isn't needed.